server/src/unifyfs_tree.c

```diff
@@ -0,0 +1,352 @@
+static unifyfs_state_filesize_t* get_state(int32_t tag)
+{
+    if (tag != glb_pmi_rank) {
+      printf("ERROR: tag %d is not expected %d\n", tag, glb_pmi_rank);  fflush(stdout);
+    }
+    return state_filesize;
+}
+
+/* function to lookup local max write offset for given file id */
+static size_t get_max_offset(int gfid)
+{
+    size_t filesize = glb_pmi_rank + 100;
+    return filesize;
+}
+
+/* given a root rank, lookup the tree structure */
+static unifyfs_tree_t* get_tree(int root)
+{
+    return &glb_tree;
+}
+
+static int rpc_invoke_filesize_request(int rank, unifyfs_state_filesize_t* st)
+{
+    int rc = (int)UNIFYFS_SUCCESS;
+
+    /* get address for specified server rank */
+    hg_addr_t addr = glb_servers[rank].margo_svr_addr;
+
+    /* get handle to rpc function */
+    hg_handle_t handle;
+    hg_return_t hret = margo_create(unifyfsd_rpc_context->svr_mid, addr,
+        unifyfsd_rpc_context->rpcs.filesize_request_id, &handle);
+    assert(hret == HG_SUCCESS);
+
+    /* fill in input struct */
+    filesize_request_in_t in;
+    in.root = (int32_t)st->root;
+    in.tag  = (int32_t)st->tag;
+    in.gfid = (int32_t)st->gfid;
+
+    /* call rpc function */
+    hret = margo_forward(handle, &in);
+    assert(hret == HG_SUCCESS);
+
+    /* wait for rpc output */
+    filesize_request_out_t out;
+    hret = margo_get_output(handle, &out);
+    assert(hret == HG_SUCCESS);
+
+    /* decode response */
+    rc = (int) out.ret;
+
+    /* free resources */
+    margo_free_output(handle, &out);
+    margo_destroy(handle);
+
+    return rc;
+}
+
+static int rpc_invoke_filesize_response(int rank, unifyfs_state_filesize_t* st)
+{
+    int rc = (int)UNIFYFS_SUCCESS;
+
+    /* get address for specified server rank */
+    hg_addr_t addr = glb_servers[rank].margo_svr_addr;
+
+    /* get handle to rpc function */
+    hg_handle_t handle;
+    hg_return_t hret = margo_create(unifyfsd_rpc_context->svr_mid, addr,
+        unifyfsd_rpc_context->rpcs.filesize_response_id, &handle);
+    assert(hret == HG_SUCCESS);
+
+    /* fill in input struct */
+    filesize_response_in_t in;
+    in.tag      = (int32_t)st->parent_tag;
+    in.filesize = (hg_size_t)st->filesize;
+
+    /* call rpc function */
+    hret = margo_forward(handle, &in);
+    assert(hret == HG_SUCCESS);
+
+    /* wait for rpc output */
+    filesize_response_out_t out;
+    hret = margo_get_output(handle, &out);
+    assert(hret == HG_SUCCESS);
+
+    /* decode response */
+    rc = (int) out.ret;
+
+    /* free resources */
+    margo_free_output(handle, &out);
+    margo_destroy(handle);
+
+    return rc;
+}
+
+unifyfs_state_filesize_t* alloc_state_filesize(void)
+{
+    state_filesize = (unifyfs_state_filesize_t*) malloc(sizeof(unifyfs_state_filesize_t));
+    return state_filesize;
+}
+
+static void filesize_response_forward(unifyfs_state_filesize_t* st)
+{
+    printf("%d: BUCKEYES response_forward\n", glb_pmi_rank);  fflush(stdout);
```

Comment:
I think we need to find a way to get a "BUCKEYES" printf in the final product ;)

Reply:
Yeah, keen eye :-). Needed someway to find my new debug messages, and for some reason that name stands out to me.  O-H!